✨ Update operator proposal (#5535)
kubernetes-sigs/cluster-api: docs/proposals/20201020-capi-provider-operator.md

operator to empower admins to handle the lifecycle of providers within the
management cluster.

> The operator will be developed in a separate repository and will have it's own release cycle. 

Reviewer: ```suggestion
The operator is developed in a separate repository [TBD] and will have its own release cycle. 
```

> The operator will be developed in a separate repository and will have it's own release cycle. 
> Possibility of moving it in core repository as a sub-project will be evaluated later.

Reviewer 2: ```suggestion
The operator will be developed in a separate repository and will have its own release cycle. 
The possibility of moving it into core repository as a sub-project will be evaluated later.
```

> The possibility of moving it into core repository as a sub-project will be evaluated later.

Reviewer: ```suggestion
```

I'd remove this for now

Reviewer: We can, although it doesn't really add more information to the users that's why I was suggesting to remove it

the operator for the functions it can. As stated in the Goals/Non-Goals, the
move operation will not be driven by the operator but rather remain within the
CLI for now. However, this is an implementation detail and will not affect the
users. The move operation and all other `clusterctl` refactoring will be

> done after core operator functionality will be implemented.

Reviewer 2: ```suggestion
done after core operator functionality is implemented.
```

##### Operator Installation

> - During the first phase of implementation `clusterctl` operations will not be 
>   supported and admin will have to install the operator manually, or in case if
>   the admin does not want to use clusterctl to install the operator, it is

Reviewer 3: ```suggestion
- During the first phase of implementation `clusterctl` won't provide support
  for managing the operator, so the admin will have to install it manually using 
  `kubectl apply` (or similar solutions); the operator yaml that will be published in the
   operator subproject release artifacts.
```
Proposing rephrased version

  possible to `kubectl apply` the operator yaml that will be published in the
  cluster-api release artifacts.
- In future `clusterctl init` will install the operator and its corresponding CRDs 
  as a pre-requisite if the operator doesn’t already exist. Please note that this
  command will consider image overrides defined in the local clusterctl config
  file.

##### Operator Upgrade
- During the first phase of implementation `clusterctl` operations will not be 
  supported and admin will have to install the operator manually, or in case if

>   the admin doesn’t want to use clusterctl, they can use kubectl apply with

Reviewer 3: Suggesting a rephrase similar to above (in case if the admin doesn’t want to use clusterctl does not make sense to me at this stage, also we are not going to publish operator artefacts in main CAPI release artefacts given that this requires release coordination and this can slow down both sides), with the difference that we are talking about upgrade.

Reviewer 3: Also, doe you have any consideration about transitioning from manually managed operator to clusterctl managed operation? It will be great if we make it clear if this is something we want to support, and possible make it easier and smooth for the users, otherwise this will require additional upgrade instructions we should plan for (or at lease document as a TODO)

Author: You are right about upgrade instructions, I will document them as TODO for now.

  the latest version of the operator yaml that will be published in the
  cluster-api release artifacts.
- In future the admin will be able to use `clusterctl upgrade operator` to 
  upgrade the operator components. Please note that this command will consider 
  image overrides defined in the local clusterctl config file. Other commands 
  such as `clusterctl upgrade apply` will also allow to upgrade the operator.
- `clusterctl upgrade plan` will identify when the operator can be upgraded by
  checking the cluster-api release artifacts.
- clusterctl will require a matching operator version. In the future, when
  clusterctl move to beta/GA, we will reconsider supporting version skew
  between clusterctl and the operator.

##### Operator Delete
- During the first phase of implementation `clusterctl` operations will not be 
  supported and admin will have to install the operator manually, or in case if

>   the admin doesn’t want to use clusterctl, they can use kubectl delete.

Reviewer 3: Same as above.
Also delete could be not straight forward given that I assume the operator will have many objects.

Author: I agree, it's written below that admins are responsible for verifying that there are no providers running in the management cluster.

  supported and admin will have to install the operator manually, or in case if
  the admin doesn’t want to use clusterctl, they can use `kubectl apply`(or similar solutions) 
  with the latest version of the operator yaml that will be published in the
  operator subproject release artifacts.
- The transition between manually managed operator and clusterctl managed
  operator will be documented later as we progress with the implementation.
- In future the admin will be able to use `clusterctl upgrade operator` to 
  upgrade the operator components. Please note that this command will consider 
  image overrides defined in the local clusterctl config file. Other commands 
  such as `clusterctl upgrade apply` will also allow to upgrade the operator.
- `clusterctl upgrade plan` will identify when the operator can be upgraded by
  checking the cluster-api release artifacts.
- clusterctl will require a matching operator version. In the future, when
  clusterctl move to beta/GA, we will reconsider supporting version skew
  between clusterctl and the operator.

##### Operator Delete
- During the first phase of implementation `clusterctl` operations will not be 

>   supported and admin will have to delete the operator manually, or in case if
>   the admin doesn’t want to use clusterctl, they can use `kubectl delete`(or similar solutions).

Reviewer 3: ```suggestion
  supported and admin will have to delete the operator manually using `kubectl delete`(or similar solutions).
```
